### PR TITLE
Introduce ser::Impossible (fixes #694)

### LINKED ISSUE #694: Provide an Impossible compound type serializer
For formats that cannot serialize one of the compound types. JSON and #691 both have needed to do this. Maybe something like:

```rust
enum Void {}
pub struct Impossible<Ok, Error>(Void, PhantomData<Ok>, PhantomData<Error>);

impl<Ok, Error> ser::SerializeSeq for Impossible<Ok, Error> {
    type Ok = Ok;
    type Error = Error;
    fn serialize_element<T: ser::Serialize>(&mut self, _: T) -> Result<(), Error> {
        match self.0 {}
    }
    fn end(self) -> Result<Ok, Error> {
        match self.0 {}
    }
}

// and the other ones
```

Then the Serializer can just do:

```rust
type SerializeSeq = Impossible<(), Error>;
```

And these can be the defaults once we have associated type defaults.

### PR DESCRIPTION
**Pull request by nox**

### CODE REVIEW

File: serde/src/ser/impossible.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+///                      len: Option<usize>)
+///                      -> Result<Self::SerializeSeq, Error> {
+///         // Given ImpossibleSerializer cannot be instantiated, the only
+///         // thing we can do here is to return an error.
+///         Err(...)
+///     }
+///
+///     ...
+/// }
+///
```
Comment (dtolnay) [inline serde/src/ser/impossible.rs]: Missing the closing ```.

File: serde/src/lib.rs [hunk 2]
```diff
@@ -86,6 +86,8 @@ mod core {
     pub use actual_core::nonzero;
 }
 
+extern crate void;
```
Comment (dtolnay) [inline serde/src/lib.rs]: I don't think we are getting enough bang for our buck out of this dependency. The entire functionality we need is provided equally well by:

```rust
enum Void {}
```

```rust
match self.void {}
```

If this were part of serde's public API then I would be all for using the void crate as a well-established way of signaling impossibility, but here it is just a private implementation detail...

File: serde/Cargo.toml [hunk 3]
```diff
@@ -18,14 +18,15 @@ travis-ci = { repository = "serde-rs/serde" }
 [features]
 default = ["std"]
 
-std = []
+std = ["void/std"]
```
Comment (nox) [inline serde/Cargo.toml]: Removed the dependency.

File: serde/src/lib.rs (hunk 2 above)
Comment (nox) [inline serde/src/lib.rs]: Removed the dependency.

File: serde/src/ser/impossible.rs (hunk 1 above)
Comment (nox) [inline serde/src/ser/impossible.rs]: Added it.

File: serde/src/ser/impossible.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+///     fn serialize_seq(self, len: Option<usize>) -> Result<(), Error> {
+///         // Given ImpossibleSerializer cannot be instantiated, the only
+///         // thing we can do here is to return an error.
+///         Err(...)
+///     }
+///
+///     ...
+/// }
+///
+pub struct ImpossibleSerializer<Ok, E> {
```
Comment (nox) [inline serde/src/ser/impossible.rs]: Renamed it.

File: serde/src/ser/mod.rs [hunk 5]
```diff
@@ -104,6 +104,7 @@ use core::cell::RefCell;
 use core::fmt::Display;
 
 mod impls;
+pub mod impossible;
```
Comment (nox) [inline serde/src/ser/mod.rs]: Reexported it from `serde::ser`.

**Comment by dtolnay** [comment]
Thanks! I released this in serde 0.9.4 and upgraded serde_json in https://github.com/serde-rs/json/commit/cdf063aefaa9fa3927b4259a94bc16fa53d84232.

**Pull request merged**